### ISSUE TITLE
"Circular dependency between the following tasks" when running tasks in a module using KSP

### ISSUE THREAD

**Issue by WhosNickDoglio** [cortinico/ktfmt-gradle#120]
## 🐛 Describe the bug
When running `ktfmtCheck` or `ktfmtFormat` in a module using KSP I get the following error 

```
Circular dependency between the following tasks:
:classes
+--- :compileJava
|    \--- :compileKotlin
|         +--- :kspKotlin
|         |    +--- :ktfmtFormatGeneratedByKspKotlin
|         |    |    \--- :kspKotlin (*)
|         |    \--- :ktfmtFormatGeneratedByKspTestKotlin
|         |         \--- :kspTestKotlin
|         |              +--- :classes (*)
|         |              +--- :compileKotlin (*)
|         |              +--- :ktfmtFormatGeneratedByKspKotlin (*)
|         |              \--- :ktfmtFormatGeneratedByKspTestKotlin (*)
|         +--- :ktfmtFormatGeneratedByKspKotlin (*)
|         \--- :ktfmtFormatGeneratedByKspTestKotlin (*)
\--- :processResources
     \--- :kspKotlin (*)
```

## ⚠️ Current behavior
Getting above mentioned error when trying to run `ktfmtCheck` or `ktfmtFormat`

## ✅ Expected behavior
Code should either be checked or formatted without circular dependency issues 

## 💣 Steps to reproduce
[Reproducer project here](https://github.com/WhosNickDoglio/ktfmt-ksp-bug)
Download the above repo and run either `ktfmtCheck` or `ktfmtFormat`

**Comment by cortinico** [comment]
Oh that's interesting 🤔  I suppose that's caused by how KSP is registering source sets that get picked up by ktfmt-gradle. Not sure why this causes a circ. dep though.

**Comment by yzhaoa** [comment]
This also happens to latest versions of the com.google.protobuf plugins, which registers the generated classes under the main sourceSet.

To work around this, you can use the following build.gradle snippet to break the dependencies:

```groovy
tasks.withType(com.ncorti.ktfmt.gradle.tasks.KtfmtBaseTask).configureEach {
  def dirs = []
  def buildDirectoryPath = project.layout.buildDirectory.asFile.get().toPath()
  source
    .matching { exclude { it.file.toPath().startsWith(buildDirectoryPath) } }
    .visit { if (it.isDirectory()) dirs.add(it.file) }
  source = files(dirs)
}
```

**Comment by cortinico** [comment]
Can we get a repro of some sort?

**Comment by WhosNickDoglio** [comment]
I linked to a reproducer in my initial post but [here it is again](https://github.com/WhosNickDoglio/ktfmt-ksp-bug).

I tried adjusting the suggested snippet to the Kotlin DSL (since that's what the reproducer project is using) but I'm still seeing the same error.

```kotlin
tasks.withType<com.ncorti.ktfmt.gradle.tasks.KtfmtBaseTask>().configureEach {
    val dirs = mutableListOf<File>()
    val buildDirectoryPath = project.layout.buildDirectory.asFile.get().toPath()
    source
        .matching { exclude { it.file.toPath().startsWith(buildDirectoryPath) } }
        .visit { if (this.isDirectory) dirs.add(this.file) }
    source = files(dirs).asFileTree
}
```

Maybe I messed up the Groovy to Kotlin translation? 🤷

**Comment by cortinico** [comment]
> I linked to a reproducer in my initial post but [here it is again](https://github.com/WhosNickDoglio/ktfmt-ksp-bug).

Thanks for sharing. I'm quite low on capacity at the moment so I won't be able to fix this till the end of the month. Any contribution is more than welcome 🙏

**Comment by cr7pt0gr4ph7** [comment]
For anybody else stumbling upon this problem, the following snippet works for me:
```kotlin
project.tasks.withType<com.ncorti.ktfmt.gradle.tasks.KtfmtBaseTask>().configureEach {
    val buildDirectoryPath = project.layout.buildDirectory.asFile.get().toPath()
    exclude { it.file.toPath().startsWith(buildDirectoryPath) }
}
```
For some reason, the default `exclude("build/")` does not actually exclude the build directory, but I don't know Gradle's internals enough to understand why.

**Comment by lukas-krecan** [comment]
I tried to fix it, but I failed. I will describe where I got so far so the others do not have to follow my steps.

The obvious idea was to exclude some Source Sets from ktfmt as it does not make much sense to run ktfmt on generated code anyway.

I tried to add `excludeSourceSets` property to `KtfmtExtension` and skip task generation for those source sets. Unfortunatelly, it does not work as the extension values are [evaluated after the plugin is applied](https://discuss.gradle.org/t/extension-initialization-vs-task-configuration/36118).
